### PR TITLE
Tighten FakeTensor reentrancy asserts, add debugging

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102091

When investigating failures in https://github.com/pytorch/pytorch/pull/100017 I realized that we were reentering FakeTensorMode even though there was already one on the stack. Although we have attempted assert for these cases in the past, e.g., as in https://github.com/pytorch/pytorch/pull/97186 it seems that the existing protections were insufficient.

In this particular case, the reapplication of FakeTensorMode was due to an interaction with NotImplemented multiple dispatch handling. If proxy tensor mode detects an unrecognized tensor type (this includes FakeTensor, if it is not tracked with a proxy), it will return NotImplemented to give this tensor a chance to unpack itself into proxyable operation. However, this is never the right thing for FakeTensor, where no unpacking is possible. However, today, FakeTensor attempts to reapply the FakeTensorMode, resulting in FakeTensorMode being twice on the stack.

This PR does a number of things:

* It adds an assert in `FakeTensorMode.__torch_dispatch__` that you must not already have this mode on the stack, this is ALWAYS an error
* It modifies `FakeTensor.__torch_dispatch__` to return `NotImplemented` if the mode is already active. This prevents us from readding the mode on the stack
* It adds a new logging artifact `not_implemented` which you can use to get debug logs about all of the times a `__torch_dispatch__` handler returned NotImplemented and why it did so. Your subclass has to manually opt into this logging, but I inserted the necessary logs for ProxyTensorMode and FakeTensor(Mode)
* `with fake_mode` now no-ops if the fake mode is already on the stack, which is what users want anyway
* I am BREAKING pre-autograd tracing, because it is currently doing something weird with the original C++ mode stack. Brian is going to follow up with a fix next week.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

**Comment by ezyang** [comment]
cc @thiagocrepaldi

**Review by thiagocrepaldi — APPROVED** [review]
LGTM

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -1098,16 +1124,34 @@ def __init__(
         # the device property
         self.in_kernel_invocation = False
 
+        self.enter_stack = []
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1127]: nit: 

` self.enter_stack : List[bool] = [] `

**Review by eellison — APPROVED** [review]
niceeeee. i do wonder whether we would be better off throwing in the case of existing fake mode. it feels a bit magical that this will be silently ignored but other modes will error in weird ways.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for fixing the tests!

**Comment by ezyang** [comment]
> niceeeee. i do wonder whether we would be better off throwing in the case of existing fake mode. it feels a bit magical that this will be silently ignored but other modes will error in weird ways.

@eellison I am throwing in this case (but indirectly via everyone returning NotImplemented)

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -252,12 +256,18 @@ def inside_mode(proxy_mode):
         proxy_mode.is_inside_mode = old
 
 def proxy_call(proxy_mode, func, pre_autograd, args, kwargs):
+    unrecognized_types = []
+
     def can_handle_tensor(x):
-        return type(x) in HANDLED_TYPES or has_proxy_slot(x, proxy_mode.tracer)
+        r = type(x) in HANDLED_TYPES or has_proxy_slot(x, proxy_mode.tracer)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:262]: I didn't like that patch which allowed for fake free variables. Maybe I could be convinced to add it again later, but in this case it was just a bug.

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -978,7 +983,13 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
...
         # subclasses of tensors to dispatch, and any FakeTensor arguments
         # will be considered eligible.
-        if any(not issubclass(t, FakeTensor) and t is not torch.Tensor for t in types):
+        unrecognized_types = [
+            t for t in types if not issubclass(t, FakeTensor) and t is not torch.Tensor
+        ]
+        if unrecognized_types:
+            not_implemented_log.debug(
+                "FakeTensor unrecognized subclass(es): %s", unrecognized_types
+            )
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:992]: (doesn't matter for this PR) should we audit all `NotImplemented` returns and add this logging in more places 🤔

**Review by bdhirsh — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:1002]: I audited all the ones on fake and proxy, which are the ones I care the most about.

**Pull request closed without merging**